Add a system service for Gifski

sunshinejr opened the pull request:
Resolves #47.

How to test it:
1. Archive the app, move it to `/Application` folder
2. In terminal make sure you have the service installed:
  a) `/System/Library/CoreServices/pbs -dump | grep Gifski.app`
  b) If you don’t see `NSBundlePath = "/Applications/Gifski.app”;`, try refreshing using `/System/Library/CoreServices/pbs -update`
3. Go to settings -> keyboard -> shortcuts -> services and enable `Gifski: Convert to GIF`

<details><summary>Sneak peak</summary>
<img src="https://user-images.githubusercontent.com/5232779/57307713-59ec7080-70e5-11e9-87dc-70ce1a404cc7.gif" />
</details>

sindresorhus commented on Gifski/Info.plist:
```diff
@@ -23,6 +23,26 @@
+			<key>NSMenuItem</key>
+			<dict>
+				<key>default</key>
+				<string>Gifski: Convert to GIF</string>
```
Any thoughts on whether we should use the above text or `Convert to GIF with Gifski`? I have a lot of services, and I see both formats being used.

sunshinejr replied:
I just used what I've seen the most in the services I use. I don't have a preference as both are used, so yeah just pick whatever you like and can change it accordingly 😄

sindresorhus replied:
I think having the app name as prefix is useful if an app exposes multiple services, but when there's only one service, `Convert to GIF with Gifski` reads better, so let's go with that.

sunshinejr replied:
Good point, gonna update the name as well 👍

sindresorhus commented on Gifski/Info.plist:
```diff
@@ -23,6 +23,26 @@
+			<dict/>
+			<key>NSSendFileTypes</key>
+			<array>
+				<string>public.movie</string>
```
We don't support all types of movies, so it should be the same as https://github.com/sindresorhus/gifski-app/blob/62877fb387299030d780cbee5df13ebb84eebf36/Gifski/Info.plist#L16-L18

sunshinejr replied:
Good catch 👍

sindresorhus commented:
Should we also set `NSServiceDescription`? It's not really clear where it would be used. Do you know?

https://developer.apple.com/library/archive/documentation/Cocoa/Conceptual/SysServices/Articles/properties.html

sindresorhus commented:
Can you document the feature in the readme and how to activate and use it? From experience, not everyone knows how Services work. And Apple has not done a very good job of making it obvious how to activate and use them.

sindresorhus commented:
Could you add a `maintaining.md` file and add the above steps for testing the service?

sindresorhus commented on Gifski/AppDelegate.swift:
```diff
@@ -63,4 +64,14 @@ final class AppDelegate: NSObject, NSApplicationDelegate {
+
+	/// This is called from NSApp as a service resolver
+	@objc
+	func convertToGif(_ pasteboard: NSPasteboard, userData: String, error: NSErrorPointer) {
```
Can you move this into an extension? Maybe we should also call it `convertToGifService`?

sunshinejr replied:
Sure, will move to extension. The name would be quite confusing for me, as this would suggest we would have some "gif service". But I see where you're coming from, you just want to indicate, in the function name, that this is called from a service right? I think we could leave the name as is with the comment and extension for now, but again up to you.

sindresorhus replied:
Alright, we don't have to change the name. It just bothers me in general how implicit all these old Cocoa APIs are.

sunshinejr commented:
@sindresorhus updated the PR with code changes and initial drafts of readme/maintaining docs. Let me know how it looks now!

Also, I didn't use `NSServiceDescription` and also don't quite understand where would it be seen. When I used the pbs dump command I didn't see any service using this property as well.

sunshinejr commented:
Also, for some reason when I used the service today (without any code changes) the app would freeze both itself and my Finder 🤔 Wonder what happened.

sindresorhus commented:
> the app would freeze both itself and my Finder 🤔 Wonder what happened.

I experienced that too. I thought it was just me doing something wrong.

kornelski commented:
If it's freezing Finder, it may be because the conversion is blocking? The callback should return ASAP and just schedule conversion for later.

sunshinejr commented:
@kornelski that's a good idea. Although it was working before and, from what I understand, the system service is opening the app for us and then we can do whatever we want in that function as the app is opened.

I did clean everything, restarted the computer (as I thought that the manual Finder opening with debug flag could cause this) and for some reason when I built and moved the app again to /Applications folder, when tapped on "Convert to GIF with Gifski" i**t opened the version of the app that I didn't even archive before**, lol (I'm pretty sure about that because it opened a version of the app with the alert as I was testing #84 for a few minutes yesterday).

So I've removed the app from /Applications, removed all previous archives, cleaned the project, did flush the pbs (`/System/Library/CoreServices/pbs -flush`) and tried the steps that I did before (basically what is in the maintaining.md) and boom - it worked. Not sure which step helped, though. Really weird.

@sindresorhus did it work for you even once and then stopped, or was it broken all the time?

sindresorhus commented on readme.md:
```diff
@@ -23,6 +23,18 @@ Requires macOS 10.13 or later.
+
+Gifski provides a [system service](https://www.computerworld.com/article/2476298/os-x-a-quick-guide-to-services-on-your-mac.html) that lets you quickly convert a video to GIF from anywhere.
+
+You first have to enable it:
```
Are you sure the user has to enable it manually? From experience, most services are auto-enabled when I install new apps.

This seems to confirm it: https://stackoverflow.com/questions/3859747/how-do-i-automatically-activate-an-item-in-the-os-x-services-menu

sunshinejr replied:
Not totally sure, I had it unchecked 100% though. We may indicate in docs that we should make sure it's enabled in the settings.

sindresorhus replied:
I have manually verified that services are enabled by default.

sindresorhus commented:
> So I've removed the app from /Applications, removed all previous archives, cleaned the project, did flush the pbs (/System/Library/CoreServices/pbs -flush) and tried the steps that I did before (basically what is in the maintaining.md) and boom - it worked. Not sure which step helped, though. Really weird.

Same. It's working now for me too. So was definitely something weird going on with different builds.

sindresorhus commented:
> Also, I didn't use NSServiceDescription and also don't quite understand where would it be seen. 

I discovered where it's used now. If you hover over a service in the Services preferences for a couple of settings, the description will be shown in a tooltip.

<img width="342" alt="Screen Shot 2019-05-22 at 12 58 10" src="https://user-images.githubusercontent.com/170270/58150585-44547a80-7c91-11e9-91ca-860b4df51368.png">
